homeassistant/components/withings/common.py

```diff
@@ -89,184 +360,272 @@ def __init__(
 
     def _request(
         self, path: str, params: Dict[str, Any], method: str = "GET"
-    ) -> Dict[str, Any]:
-        return run_coroutine_threadsafe(
-            self.async_do_request(path, params, method), self._hass.loop
-        ).result()
-
-    async def async_do_request(
-        self, path: str, params: Dict[str, Any], method: str = "GET"
     ) -> Dict[str, Any]:
         """Perform an async request."""
-        await self.session.async_ensure_token_valid()
-
-        response = await self._hass.async_add_executor_job(
-            partial(
-                requests.request,
-                method,
-                f"{self.URL}/{path}",
-                params=params,
-                headers={
-                    "Authorization": "Bearer %s"
-                    % self._config_entry.data["token"]["access_token"]
-                },
-            )
+        asyncio.run_coroutine_threadsafe(
+            self.session.async_ensure_token_valid(), self._hass.loop
         )
 
+        access_token = self._config_entry.data["token"]["access_token"]
+        response = requests.request(
+            method,
+            f"{self.URL}/{path}",
+            params=params,
+            headers={"Authorization": f"Bearer {access_token}"},
+        )
         return response.json()
 
 
+def json_message_response(message: str, message_code: int) -> Response:
+    """Produce common json output."""
+    return http.HomeAssistantView.json({"message": message, "code": message_code}, 200)
+
+
+class WebhookAvailability(IntEnum):
+    """Represents various statuses of webhook availability."""
+
+    SUCCESS = 0
+    CONNECT_ERROR = 1
+    HTTP_ERROR = 2
+    NOT_WEBHOOK = 3
+
+
+def async_is_valid_etag(etag: str) -> bool:
+    """Check that the etag came from the withings webhook."""
+    return etag is not None and etag.startswith(f"{const.WEBHOOK_TEST_ETAG}_")
+
+
+def async_create_etag() -> str:
+    """Create a new webhook etag."""
+    return f"{const.WEBHOOK_TEST_ETAG}_{time.time()}"
+
+
+async def async_test_webhook_availability(
+    hass: HomeAssistant, callback_url: str
+) -> WebhookAvailability:
+    """Test withings webhook availability."""
+    session = async_get_clientsession(hass)
+    error = None
+    error_message = None
+
+    try:
+        response = await session.head(callback_url)
```

Comment:
Because of NAT hairpinning, you might end up with a url that is accessible from outside of your network, but not internally. 

You should also add a timeout.